crates/forge-runner/src/gas.rs

```diff
@@ -1,79 +1,164 @@
+use blockifier::abi::constants;
 use std::collections::HashMap;
 
 use crate::test_case_summary::{Single, TestCaseSummary};
-use blockifier::fee::fee_utils::calculate_tx_l1_gas_usage;
-use blockifier::fee::gas_usage::calculate_tx_gas_usage;
-use blockifier::fee::os_resources::OS_RESOURCES;
-use blockifier::fee::os_usage::get_additional_os_resources;
-use blockifier::state::cached_state::{CachedState, StateChangesCount};
-use blockifier::state::errors::StateError;
-use blockifier::transaction::transaction_types::TransactionType;
-use blockifier::{
-    abi::constants, block_context::BlockContext, execution::entry_point::ExecutionResources,
-    transaction::objects::ResourcesMapping,
+use blockifier::context::TransactionContext;
+use blockifier::fee::eth_gas_constants;
+use blockifier::fee::fee_utils::calculate_tx_gas_vector;
+use blockifier::fee::gas_usage::{
+    get_consumed_message_to_l2_emissions_cost, get_da_gas_cost,
+    get_log_message_to_l1_emissions_cost,
 };
-use cairo_vm::vm::runners::cairo_runner::ExecutionResources as VmExecutionResources;
+use blockifier::state::cached_state::CachedState;
+use blockifier::state::errors::StateError;
+use blockifier::transaction::objects::{GasVector, HasRelatedFeeType, ResourcesMapping};
+use blockifier::utils::{u128_from_usize, usize_from_u128};
+use cairo_vm::vm::runners::builtin_runner::SEGMENT_ARENA_BUILTIN_NAME;
+use cairo_vm::vm::runners::cairo_runner::ExecutionResources;
 use cheatnet::runtime_extensions::call_to_blockifier_runtime_extension::rpc::UsedResources;
 use cheatnet::state::ExtendedStateReader;
 
 pub fn calculate_used_gas(
-    block_context: &BlockContext,
+    transaction_context: &TransactionContext,
     state: &mut CachedState<ExtendedStateReader>,
-    resources: &UsedResources,
+    mut resources: UsedResources,
 ) -> Result<u128, StateError> {
-    let total_vm_usage = get_total_vm_usage(&resources.execution_resources);
-    let mut state_changes = state.get_actual_state_changes_for_fee_charge(
-        block_context.fee_token_addresses.eth_fee_token_address,
-        None,
-    )?;
+    add_syscall_resources(transaction_context, &mut resources);
+
+    let l1_data_cost = get_l1_data_cost(transaction_context, state)?;
+    let messaging_gas_vector = get_messages_costs(
+        &resources.l2_to_l1_payloads_lengths,
+        &resources.l1_handler_payloads_lengths,
+    );
+
+    let l1_and_vm_costs = get_l1_and_vm_costs(
+        l1_data_cost,
+        transaction_context,
+        resources.execution_resources,
+    );
```

Comment:
[the most nit of all nits] I would swap the order of the vars, since the `l1_data_cost` is used only for `l1_and_vm_costs` xD sorry, I had to